BedrockServer.cpp

```diff
@@ -163,6 +163,8 @@ void BedrockServer::sync(const SData& args,
     // corresponding number of journal tables. "-readThreads" exists only for backwards compatibility.
     int workerThreads = args.calc("-workerThreads");
 
+    int replicationThreads = 8; // TODO: make configurable.
```

Comment:
Actually... what would you think of just trying to spawn a new thread for every single command, and see if that works?  Because there's really on reason to have 8 or any specific number of replication threads.  I'd love to see if we can just create as many threads as needed, on demand.  Because that will help us understand if that design will work for Bedrock3.

Reply:
In fact, a different way to think of this entirely isn't as a pool of workers, but rather each command *is a thread*.  So rather than the sync thread queueing `BEGIN_TRANSACTION` commands to a fixed-size pool of workers monitoring that queue, what if instead we just spawn a thread for each command?  This begins experimenting with a queue-less design philosophy right now. 

How this will work in the future
====
Eventually, in a Bedrock3 world, it'd work like this:

1. Follower A's "acceptor thread" receives a new socket from the client; it accepts it, creates a new "command thread" to process it.

2. That follower A's command thread (A1) reads the (for this example) `INSERT INTO blah VALUES (1);` query from the socket.  A1 recognizes it's a write command, and thus needs to escalate.  A1 does this by opening a socket to the leader (or, more likely, reuses one from a recycled pool somehow, but that's a potentially premature optimization), just sending the original `INSERT INTO blah VALUES (1);` query, more or less verbatim, and then blocks waiting or the response.

3. The leader L's "acceptor thread" receives the socket from the "client" (in this case, follower A's command thread A1); it accepts it, and creates a new command thread L1 to process it.

4. The leader's L1 command thread reads the query from the socket, and concludes it's a write command.  It starts a local transaction, then executes the query.  (If this command had any external networking it'd do it now, blocking the thread while waiting for the response.  But this command doesn't.)

5. L1 then opens/recycles a socket _to every single follower_ (including A and B), and sends `BEGIN_TRANSACTION`.

6. Follower B's acceptor thread receives this socket, creates command B1 to process it.

7. B1 sees it's a replication command, and:
    1. monitors the commit count to wait its turn, then 
    2. begins the transaction, then 
    3. replies _on the L1>B1 socket_ with `APPROVE_TRANSACTION`.  
    4. It then blocks waits on that same socket to see what to do next.

8. L1 is waiting on responses to its `BEGIN_TRANSACTION` commands; it gets `APPROVE_TRANSACTION` responses back from a quorum of peers.  So it:
    1. sends back _on those same sockets_ `COMMIT_TRANSACTION`
    2. responds back to A1 with `OK`
    3. self destructs

9. B1 is waiting for a response, and gets `COMMIT_TRANSACTION`.  It:
    1. commits, 
    2. closes the socket, and 
    3. self destructs.

10. A1, which has been waiting patiently for a response to its query, gets back `OK`.  It then turns around and:
    1. sends that to the client, 
    2. closes the socket, and 
    3. self destructs.

How this could work today
====
So obviously the above is a bit bigger than we would want to bite off right now.  But what if we approached it in stages, the first of which could be:

1. Everything starts just like today

2. We update the "sync thread" to just create new command threads for everything it pulls off of the peer socket, exactly as we plan in the future for Bedrock3.

3. The `BEGIN_TRANSACTION` command thread:
    1. waits for its turn, then
    2. begins the transaction, then
    3. locks the leader peer socket
    4. Sends `APPROVE_TRANSACTION` on that socket.  This isn't ideal, but for now we continue multiplexing all peer communications over the peer thread.
    5. Because the `BEGIN_TRANSACTION` command thread doesn't own the leader socket, it can't directly wait for a response from the leader.  Rather, it waits for its hash to appear in `_replicationHashesToCommit/Rollback` datastructures.
    6. Then it either commits or rolls back
    7. Finally, it self destructs

4. The `COMMIT/ROLLBACK_TRANSACTION` command threads simply:
    1. insert the hash into `_replicationHashesToCommit/Rollback`, and then
    2. self destruct

So, we can use this multi-threaded replication code to start taking baby steps to our Bedrock3 design.

Reply:
And actually... what if we view the `APPROVE_COMMIT` command _just like any other escalated command_.  And right now, we escalate all commands over the peer socket, with responses coming back as `ESCALATE_RESPONSE`.  What if we just treat `COMMIT_TRANSACTION` and `ROLLBACK_TRANSACTION` commands just like `ESCALATE_RESPONSE`.  

So rather than storing those results in `_replicationHashesToCommit/Rollback` -- which only makes sense to replication -- we just have a `map<commandID, SData> _escalateResponses` datastructure, which initially is only used for replications, but eventually would allow these worker threads to process any commands (and get back escalated responses) multiplexed over the leader peer connection.

This means rather than polling the `_replicationHashesToCommit/Rollback` datastructures to see if it should rollback/commit, the `BEGIN_TRANSACTION` command thread just waits for a response to its `APPROVE_TRANSACTION` to appear in `_escalateResponses`.  Then it looks at that response to see if it's `COMMIT_TRANSACTION` or `ROLLBACK_TRANSACTION`.

Then when we eventually expand the "command thread" framework to handle all commands, they escalate to the leader exactly like the `BEGIN_TRANSACTION` command thread (by locking the leader peer socket and sending) -- and they wait for responses the same way (by monitoring `_escalateResponses`).

And then someday we enable command threads to escalate *directly* to the leader -- without multiplexing over the leader peer socket.  And then those command threads stop monitoring `_escalateResponses`, they just use blockign IO on the socket it escalated to the leader.  Once all peer multiplexing is dead, we remove support for peer socket multiplexing (and remove the `_escalateResponses` data structure entirely).

Reply:
Incidentally, so far as I can determine, "selective synchronization" doesn't really change any of this.  All selective sync is do is send the `COMMIT_TRANSACTION` without waiting for `APPROVE_TRANSACTION`.  But none of that changes any of the above logic.  If the leader sends `BEGIN_TRANSACTION(hashX)` and `COMMIT_TRANSACTION(hashX)` back to back, all it means is that the `_escalatedResponses` datastructure will have the `COMMIT_TRANSACTION` for `hashX` ready to go the instant `BEGIN_TRANSACTION` finishes its query.  Everything works exactly the same so long as the follower is concerned (ie, it still sends `APPROVE_TRANSACTION to the leader), it's just as if the leader is instant fast in responding.

Reply:
So, I thought about several ideas along these lines as well, but the main thing preventing this is nothing in the above, but the way the parallel DB works. We can't spawn up arbitrary threads with the current DB architecture, because we need each thread to have it's own database handle out of a fixed pool, and the reasoning for that is because each DB handle needs it's own journal table to prevent conflicts.

One thing we could do is create an overabundant supply of journal tables. Say, 1024 of them, and then each thread could just grab a journal table and use that. This would be "potentially conflicting" if two threads did get the same journal table, but could be resolved with the existing conflict resolution mechanism.

We could probably assign out journal tables in a lock-free fashion pretty easily, just keep an `atomic<int>` that's the current table to assign, and each new thread increments that number. With 1024 journal tables available, any two threads that *are* assigned the same journal table should be 1024 commands apart in the order in which they were received, and 5 commands apart on a given CPU core (assuming that commands ran in uninterrupted chunks on a single core) and thus, fairly unlikely to conflict.

If we had a number of commands in progress exceeding the number of CPU cores that we have available, then it could end up processing individual commands more slowly, though overall throughput should remain the same. I wonder how often this would actually happen, though.

So, now that I thought of a potential journal table solution, I'm more excited about this idea again.